internal/sim/api.go

```diff
@@ -249,36 +306,84 @@ func validateWorkload(t reflect.Type) error {
 }
 
 // Run runs simulations for the provided duration.
-func (s *Simulator) Run(duration time.Duration, opts Options) Results {
-	// TODO(mwhittaker): Implement.
-	return Results{}
+func (s *Simulator) Run(duration time.Duration) (Results, error) {
+	// TODO(mwhittaker): Run many simulations with many different
+	// hyperparameters in many different goroutines. For now, we just run one.
+	// TODO(mwhittaker): Read and run graveyard entries.
+	ctx, cancel := context.WithTimeout(context.Background(), duration)
+	defer cancel()
+	opts := options{
+		Seed:        time.Now().UnixNano(),
+		NumOps:      10,
+		NumReplicas: 1,
+		FailureRate: 0.1,
+		YieldRate:   0.25,
+	}
+	return s.runOne(ctx, opts)
 }
 
-// A registrar is the canonical implementation of Registrar.
-//
-// TODO(mwhittaker): For now, the registrar just validates registered fakes and
-// generators. In the future, it will do more than that.
-type registrar struct {
-	t          *testing.T          // underlying test
-	w          reflect.Type        // workload type
-	generators map[string]struct{} // methods with registered generators
+// runOne runs a single simulation.
+func (s *Simulator) runOne(ctx context.Context, opts options) (Results, error) {
+	// Construct an instance of the workload struct.
+	workload := reflect.New(s.w.Elem()).Interface().(Workload)
+
+	// Call the struct's Init method.
+	r := &registrar{t: s.t, w: s.w}
+	if err := workload.Init(r); err != nil {
+		return Results{}, err
+	}
+
+	// Simulate!
+	sim, err := newSimulator(s.t.Name(), workload, s.regsByIntf, s.config, r.fakes, r.ops(), opts)
+	if err != nil {
+		return Results{}, err
+	}
+	return sim.Simulate(ctx)
+}
+
+// validatingRegistrar is a Registrar that validates registered fakes and
+// generators.
+type validatingRegistrar struct {
+	t          *testing.T                             // underlying test
+	w          reflect.Type                           // workload type
+	regsByIntf map[reflect.Type]*codegen.Registration // registered components, by interface
+	fakes      map[reflect.Type]struct{}              // registered fakes, by interface
+	generators map[string]struct{}                    // registered generators, by method name
 }
 
 // RegisterFake implements the Registrar interface.
-func (r *registrar) RegisterFake(weavertest.FakeComponent) {
-	// TODO(mwhittaker): Implement.
+func (r *validatingRegistrar) RegisterFake(fake FakeComponent) {
```

Comment:
I presume you had to split it this way because of cached validation?

Reply:
Yeah. I merged the two registrar implementations now.